atlasdb-cassandra/src/test/java/com/palantir/atlasdb/cassandra/ReloadingCloseableContainerTest.java

```diff
@@ -0,0 +1,325 @@
+        ReloadingCloseableContainer<Closeable> container = getContainer();
+        assertThatThrownBy(() -> container.runWithResource(_closeable -> {
+                    throw new SafeIllegalStateException("Test");
+                }))
+                .isInstanceOf(SafeIllegalStateException.class);
+    }
+
+    @Test
+    public void runWithDoesNotBlockResourceUpdates() {
+        ReloadingCloseableContainer<Closeable> container = getContainer();
+        CyclicBarrier ensureInitialExecuteHasStartedBarrier = new CyclicBarrier(2);
+        CyclicBarrier manuallyFinishRunWithBarrier = new CyclicBarrier(2);
+        executor.execute(() -> container.executeWithResource(closeable -> {
+            awaitOrFail(ensureInitialExecuteHasStartedBarrier);
+            awaitOrFail(manuallyFinishRunWithBarrier);
+            assertThat(closeable).isEqualTo(mockCloseable);
+        }));
+
+        awaitOrFail(ensureInitialExecuteHasStartedBarrier);
+        // Map and subscriber updates happen sequentially, but on the same thread. Therefore, the next call won't
+        // complete until after the execute above completes, even if the map itself has finished.
+        executor.execute(() -> settableRefreshable.update(UPDATED_VALUE));
+
+        // The earlier executeWithResource call cannot complete yet, and so if the update was blocked, this following
+        // assertion would fail.
+
+        // It is possible for this to fail, if this call ends up scheduled before the update has started...
+        Future<?> executeWithResourceFuture = executor.submit(() ->
+                container.executeWithResource(closeable -> assertThat(closeable).isEqualTo(mockCloseableExtra)));
+
+        Awaitility.await().atMost(Duration.ofSeconds(1)).until(executeWithResourceFuture::isDone);
+        awaitOrFail(manuallyFinishRunWithBarrier);
+    }
+
+    private void triggerEvent(ReloadingCloseableContainer<Closeable> container, AtomicInteger refreshableValue) {
+        double random = ThreadLocalRandom.current().nextDouble();
+        if (random < 0.2) {
+            settableRefreshable.update(refreshableValue.incrementAndGet());
+        } else {
+            container.executeWithResource(_closeable -> {});
+        }
+    }
+
+    private CompletableFuture<?> runManyTimesAndGetSingleFuture(int numberOfRuns, Runnable task) {
+        CompletableFuture<?>[] allFutures = IntStream.range(0, numberOfRuns)
+                .mapToObj(_id -> CompletableFuture.runAsync(task, executor))
+                .toArray(CompletableFuture[]::new);
+        return CompletableFuture.anyOf(allFutures);
+    }
+
+    private ReloadingCloseableContainer<Closeable> getContainer() {
+        return ReloadingCloseableContainer.of(closeableFactory, settableRefreshable);
+    }
+
+    private static void awaitOrFail(CyclicBarrier barrier) {
+        try {
+            barrier.await();
+        } catch (InterruptedException | BrokenBarrierException e) {
```

Comment:
it's standard when catching an `InterruptedException` to do `Thread.currentThread().interrupt()`